File: weapon_management/weapon_management/weapon_management/report/weapon_inventory/weapon_inventory.js


frappe.query_reports["Weapon Inventory"] = {
    "filters": [
        {
            "fieldname": "unit",
            "label": __("Unit"),
            "fieldtype": "Link",
            "options": "Unit Master",
            "reqd": 0
        },
        {
            "fieldname": "weapon",
            "label": __("Weapon"),
            "fieldtype": "Link",
            "options": "Weapon Master",
            "reqd": 0
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "reqd": 0
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "reqd": 0
        },
        {
            "fieldname": "Clear Filters",
            "label": __("Clear Filters"),
            "fieldtype": "Button",
            "reqd": 0,
            "click": function() {
                frappe.query_report.set_filter_value("unit", "");
                frappe.query_report.set_filter_value("weapon", "");
                frappe.query_report.set_filter_value("from_date", "");
                frappe.query_report.set_filter_value("to_date", "");
                frappe.query_report.refresh();
            },
        }
    ]
};

File: weapon_management/weapon_management/weapon_management/report/weapon_inventory/weapon_inventory.py

import frappe

def execute(filters=None):

    columns = ["Weapon", "No Of Weapon"]

    data = []

    filters_dict = frappe._dict(filters)

    query = """
        SELECT 
            twid.weapon_name AS "Weapon",
            COUNT(twid.weapon_name) AS "No Of Weapon"
        FROM "tabWeapon In Details" twid
        INNER JOIN "tabUnit Master" tum ON twid.unit_location = tum.name
        INNER JOIN "tabWeapon Category Master" twcm ON twid.weapon_category = twcm.name
    """

    where_clause = "WHERE 1=1" 

    if filters_dict.get("unit"):
        where_clause += f" AND tum.name = %(unit)s"
    
    if filters_dict.get("weapon"):
        where_clause += f" AND twid.weapon_name = %(weapon)s"

    if filters_dict.get("from_date") and filters_dict.get("to_date"):
        where_clause += " AND twid.date_acquired BETWEEN %(from_date)s AND %(to_date)s"

    query += where_clause

    query += " GROUP BY twid.weapon_name;"

    result = frappe.db.sql(query, filters_dict, as_dict=True)

    for row in result:
        data_row = [row.get("Weapon"), row.get("No Of Weapon")]
        data.append(data_row)

    return columns, data
